// social-media/dist/scheduler/event.model.js
"use strict";
var EventModel = (function () {
    function EventModel() {
    }
    EventModel.prototype.toEventData = function (group) {
        var event = this;
        var oneDay = 24 * 60 * 60 * 1000;
        var end;
        var start = new UTC(event.startTime.getTime());
        if (event.allDay) {
            var diffDays = Math.round(Math.abs((event.endTime.getTime() - event.startTime.getTime()) / (oneDay)));
            end = new UTC(event.startTime.getTime());
            end.addDays(diffDays);
        }
        else {
            end = new UTC(event.endTime.getTime());
        }
        var newEvent = {
            name: event.name,
            eventId: event.id,
            groupId: (group != null) ? group.groupId : null,
            group: (group != null) ? {
                groupId: group.groupId,
                name: group.name
            } : null,
            location: event.location,
            instructor: event.instructor,
            startTime: start.getDateObj(),
            endTime: end.getDateObj(),
            description: event.description,
            allDay: event.allDay,
            repeatEvent: event.repeatEvent
        };
        return newEvent;
    };
    EventModel.prototype.copyFrom = function (event) {
        this.id = event.id;
        this.name = event.name;
        this.instructor = event.instructor;
        this.startTime = event.startTime;
        this.endTime = event.endTime;
        this.description = event.description;
        this.allDay = event.allDay;
        this.repeatEvent = event.repeatEvent;
        this.groupId = event.groupId;
        this.location = event.location;
    };
    return EventModel;
}());
exports.EventModel = EventModel;
